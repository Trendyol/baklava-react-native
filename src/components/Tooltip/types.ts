import { VariantProps } from '@ergenekonyigit/restyle';
import { Theme } from '../../theme';
import { View, ViewStyle } from 'react-native';

export type Position = 'top' | 'bottom';

export type TooltipProps = React.PropsWithChildren<{
  id: string;
  content: string;
  position: Position;
  overlay?: boolean;
  hole?: boolean;
  holePadding?: number;
  pressable?: boolean;
  color?: VariantProps<Theme, 'colors'>['variant'];
  wrapperStyle?: ViewStyle;
  onClose?: (id: string) => void;
}>;

export type TooltipRef = {
  show: () => void;
  hide: () => void;
};

export type TooltipData = {
  id: string;
  position: Position;
  viewport: Viewport;
  content: string;
  color: TooltipProps['color'];
  overlay: boolean;
  hole: boolean;
  holePadding: number;
  onClose: TooltipProps['onClose'];
  triggerRef: React.RefObject<View>;
};

export interface TooltipContextType {
  activeTooltip: TooltipData;
  overlayViewport?: Viewport;
  show: (tooltipData: Omit<TooltipData, 'viewport'>) => void;
  hide: (key: string) => void;
  clearAll: () => void;
}

export interface Viewport {
  x: number;
  y: number;
  w: number;
  h: number;
  px: number;
  py: number;
}

export type Calculation = {
  height: number;
  top: number;
  pinX: number;
  pinY: number;
  pinYOverflow: number;
  pinXOverflow: number;
};
