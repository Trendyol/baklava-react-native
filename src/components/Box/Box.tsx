import { createBox } from '@ergenekonyigit/restyle';
import { Theme } from '../../theme';

const Box = createBox<Theme>();

export default Box;
