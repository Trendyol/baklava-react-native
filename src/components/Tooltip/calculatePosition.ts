import { Dimensions } from 'react-native';
import theme from '../../theme';
import { deviceWidth } from '../../utils/dimentions';
import { Calculation, TooltipData, Viewport } from './types';

export const PIN_SIZE = 10;
const PIN_HYPOT = Math.hypot(PIN_SIZE, PIN_SIZE);
const PIN_HEIGHT = PIN_HYPOT / 2;
const PIN_MARGIN = 4;

export function calculatePosition(
  tooltipData: TooltipData,
  overlayViewport?: Viewport,
): Calculation {
  switch (tooltipData.position) {
    case 'top':
      return calculateTopPosition(tooltipData, overlayViewport);
    case 'bottom':
      return calculateBottomPosition(tooltipData, overlayViewport);
  }
}

function calculateTopPosition(data: TooltipData, overlayViewport?: Viewport) {
  const hasHole = data.overlay && data.hole;
  return {
    height:
      data.viewport.py -
      PIN_HEIGHT -
      PIN_MARGIN -
      (hasHole ? data.holePadding : 0) -
      (overlayViewport?.py ?? 0),
    top: 0,
    pinX: Math.max(
      Math.min(
        data.viewport.px + data.viewport.w / 2 - PIN_HEIGHT,
        deviceWidth - theme.spacing.m * 2 - theme.borderRadii.m / 2,
      ),
      Math.ceil(theme.spacing.m + theme.borderRadii.m / 2),
    ),
    pinY: PIN_HEIGHT,
    pinYOverflow: -(PIN_HYPOT - PIN_SIZE) / 2,
    pinXOverflow: (PIN_HYPOT - PIN_SIZE) / 2,
  };
}

function calculateBottomPosition(
  data: TooltipData,
  overlayViewport?: Viewport,
) {
  const hasHole = data.overlay && data.hole;
  return {
    height:
      Dimensions.get('screen').height -
      data.viewport.py -
      data.viewport.h -
      PIN_HEIGHT -
      PIN_MARGIN,
    top:
      data.viewport.py +
      data.viewport.h +
      PIN_HEIGHT +
      PIN_MARGIN +
      (hasHole ? data.holePadding : 0) -
      (overlayViewport?.py ?? 0),
    pinX: Math.max(
      Math.min(
        data.viewport.px + data.viewport.w / 2 - PIN_HEIGHT,
        deviceWidth - theme.spacing.m * 2 - theme.borderRadii.m / 2,
      ),
      Math.ceil(theme.spacing.m + theme.borderRadii.m / 2),
    ),
    pinY: -PIN_HEIGHT,
    pinYOverflow: (PIN_HYPOT - PIN_SIZE) / 2,
    pinXOverflow: (PIN_HYPOT - PIN_SIZE) / 2,
  };
}
