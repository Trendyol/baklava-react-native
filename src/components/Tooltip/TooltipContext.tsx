import React, { createContext, ReactNode, useContext, useState } from 'react';
import { Dimensions, TouchableWithoutFeedback, View } from 'react-native';
import Svg, { Defs, G, Mask, Rect } from 'react-native-svg';
import Box from '../Box/Box';
import TooltipView from './TooltipView';
import { Viewport, TooltipContextType, TooltipData } from './types';
import theme from '../../theme';

export const TooltipContext = createContext<TooltipContextType | undefined>(
  undefined,
);

export const TooltipProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const overlayRef = React.useRef<View>(null);
  const [viewport, setViewport] = React.useState<Viewport>();
  const [stack, setStack] = useState<TooltipData[]>([]);

  const activeTooltip = React.useMemo(() => {
    return stack[0];
  }, [stack]);

  const onLayout = React.useCallback(() => {
    overlayRef.current?.measure((x, y, w, h, px, py) => {
      setViewport({
        x,
        y,
        w,
        h,
        px,
        py,
      });
    });
  }, []);

  const addToStack = React.useCallback(
    (tooltipData: Omit<TooltipData, 'viewport'>) => {
      if (tooltipData) {
        tooltipData.triggerRef.current?.measure?.((x, y, w, h, px, py) => {
          if (x ?? w ?? px === undefined) {
            tooltipData.onError?.(tooltipData.id);
            return;
          }
          const _viewport: Viewport = {
            x,
            y,
            w,
            h,
            px,
            py,
          };
          setStack(s => [...s, { ...tooltipData, viewport: _viewport }]);
        });
      }
    },
    [],
  );

  const removeFromStack = React.useCallback(
    (id: string) => {
      const temp = stack.find(t => t.id === id);
      if (temp) {
        setStack(stack.filter(t => t.id !== temp?.id));
        temp?.onClose?.(id);
      }
    },
    [stack],
  );

  const removeAllStack = React.useCallback(() => {
    setStack([]);
  }, []);

  const removeActiveTooltip = React.useCallback(() => {
    removeFromStack(activeTooltip?.id);
  }, [activeTooltip, removeFromStack]);

  const render = React.useMemo(() => {
    if (!activeTooltip) {
      return null;
    }

    const { id, overlay, hole, highlight } = activeTooltip;
    return (
      <>
        <TouchableWithoutFeedback
          onPress={removeActiveTooltip}
          accessibilityLabel={`tooltip-overlay-wrapper-${id}`}
          testID={`tooltip-overlay-wrapper-${id}`}>
          <Box
            position="absolute"
            flex={1}
            height={'100%'}
            width={'100%'}
            onLayout={onLayout}
            ref={overlayRef}>
            {overlay && hole && highlight ? (
              <Box
                testID={`tooltip-highlight-${activeTooltip.id}`}
                bg="white"
                position="absolute"
                borderRadius="m"
                justifyContent="center"
                alignItems="center"
                top={
                  activeTooltip.viewport.py -
                  (viewport?.py ?? 0) -
                  activeTooltip.holePadding
                }
                left={activeTooltip.viewport.px - activeTooltip.holePadding}
                width={activeTooltip.viewport.w + activeTooltip.holePadding * 2}
                height={
                  activeTooltip.viewport.h + activeTooltip.holePadding * 2
                }>
                {activeTooltip.children}
              </Box>
            ) : null}
            <Svg height="100%" width="100%" pointerEvents="none">
              {overlay ? (
                <G>
                  <Defs>
                    <Mask id="clip">
                      <Rect fill="#ffffff" width="100%" height="100%" />
                      {/* hole */}
                      {viewport && hole ? (
                        <Rect
                          accessibilityLabel={`tooltip-hole-${id}`}
                          testID={`tooltip-hole-${id}`}
                          fill="#000"
                          ry={6}
                          rx={6}
                          x={
                            activeTooltip.viewport.px -
                            activeTooltip.holePadding
                          }
                          y={
                            activeTooltip.viewport.py -
                            activeTooltip.holePadding -
                            (viewport?.py ?? 0)
                          }
                          width={
                            activeTooltip.viewport.w +
                            activeTooltip.holePadding * 2
                          }
                          height={
                            activeTooltip.viewport.h +
                            activeTooltip.holePadding * 2
                          }
                        />
                      ) : undefined}
                    </Mask>
                  </Defs>
                  <Rect
                    accessibilityLabel={`tooltip-overlay-${id}`}
                    testID={`tooltip-overlay-${id}`}
                    width={Dimensions.get('screen').width}
                    height={Dimensions.get('screen').height}
                    clipRule={'evenodd'}
                    fillRule={'nonzero'}
                    vectorEffect="inherit"
                    fill={theme.colors.neutralDarker}
                    fillOpacity=".7"
                    mask={hole ? 'url(#clip)' : undefined}
                  />
                </G>
              ) : undefined}
            </Svg>
          </Box>
        </TouchableWithoutFeedback>
        <TooltipView />
      </>
    );
  }, [activeTooltip, onLayout, removeActiveTooltip, viewport]);

  const value = React.useMemo<TooltipContextType>(
    () => ({
      activeTooltip,
      overlayViewport: viewport,
      show: addToStack,
      hide: removeFromStack,
      clearAll: removeAllStack,
    }),
    [activeTooltip, addToStack, removeAllStack, removeFromStack, viewport],
  );

  return (
    <TooltipContext.Provider value={value}>
      {children}
      {render}
    </TooltipContext.Provider>
  );
};

export const useTooltipContext = () => {
  const context = useContext(TooltipContext);
  if (!context) {
    throw new Error('useTooltipContext must be used within a TooltipProvider');
  }
  return context;
};
