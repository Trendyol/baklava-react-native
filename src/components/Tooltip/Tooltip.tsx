import React from 'react';
import { TouchableOpacity, View } from 'react-native';
import Box from '../Box/Box';
import { useTooltipContext } from './TooltipContext';
import { TooltipData, TooltipProps, TooltipRef } from './types';

const Tooltip = React.forwardRef<TooltipRef, TooltipProps>(
  (
    {
      children,
      id,
      pressable = true,
      content,
      overlay = false,
      hole = false,
      highlight = false,
      position = 'top',
      wrapperStyle,
      holePadding = 4,
      color = 'infoKey',
      onClose = () => null,
      onError = () => null,
    },
    ref,
  ) => {
    const { show, hide } = useTooltipContext();
    const triggerRef = React.useRef<View>(null);
    const [tooltipData, setTooltipData] =
      React.useState<Omit<TooltipData, 'viewport'>>();
    const isPending = React.useRef(false);

    React.useImperativeHandle(
      ref,
      () => {
        return {
          show: () => {
            if (tooltipData) {
              show(tooltipData);
            } else {
              isPending.current = true;
            }
          },
          hide: () => {
            if (tooltipData) {
              hide(tooltipData?.id);
            }
          },
        };
      },
      [tooltipData, show, hide],
    );

    const onLayout = React.useCallback(() => {
      const data: Omit<TooltipData, 'viewport'> = {
        id,
        position,
        content,
        overlay,
        hole,
        holePadding,
        highlight,
        color,
        onClose,
        onError,
        triggerRef,
        children,
      };
      setTooltipData(data);
      if (isPending.current) {
        show(data);
      }
    }, [
      id,
      position,
      content,
      overlay,
      hole,
      holePadding,
      highlight,
      color,
      onClose,
      onError,
      show,
      children,
    ]);

    const openTooltip = React.useCallback(() => {
      if (tooltipData) {
        show(tooltipData);
      }
    }, [tooltipData, show]);

    return (
      <TouchableOpacity
        testID={`tooltip-trigger-${id}`}
        accessibilityLabel={`tooltip-trigger-${id}`}
        disabled={!pressable}
        onPress={openTooltip}
        style={wrapperStyle}>
        <Box
          testID={`tooltip-child-wrapper-${id}`}
          ref={triggerRef}
          onLayout={onLayout}>
          {children}
        </Box>
      </TouchableOpacity>
    );
  },
);

export default Tooltip;
