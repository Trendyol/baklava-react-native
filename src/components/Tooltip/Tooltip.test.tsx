import React from 'react';
import { cleanup, fireEvent, render, waitFor } from '../../test-utils';
import Text from '../Text/Text';
import Tooltip from './Tooltip';
import { TooltipProvider, useTooltipContext } from './TooltipContext';
import Box from '../Box/Box';
import Button from '../Button/Button';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { TooltipRef } from './types';
import { deviceHeight } from '../../utils/dimentions';

describe('Tooltip', () => {
  const mockOverlayMeasure = jest.fn();
  const mockChildWrapperTopMeasure = jest.fn();
  const mockChildWrapperBottomMeasure = jest.fn();
  const mockChildWrapperOnLayoutParam = {
    nativeEvent: {
      layout: {
        width: 200,
        height: 30,
      },
    },
  };

  beforeEach(() => {
    mockChildWrapperTopMeasure.mockImplementation(f => {
      // x,y,w,h,px,py
      f(0, 0, 200, 30, 16, 16);
    });
    mockChildWrapperBottomMeasure.mockImplementation(f => {
      // x,y,w,h,px,py
      f(0, 0, 200, 30, 16, 800);
    });
    mockOverlayMeasure.mockImplementation(f => {
      // x,y,w,h,px,py
      f(0, 0, 400, 800, 0, 62);
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
    cleanup();
  });
  test('should render top correctly', async () => {
    // when
    const { toJSON } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            content={'content Lorem ipsum dolor sit amet'}
            position={'top'}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render bottom correctly', async () => {
    // when
    const { toJSON } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            content={'content Lorem ipsum dolor sit amet'}
            position={'bottom'}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    expect(toJSON()).toMatchSnapshot();
  });

  test('should not render without provider', async () => {
    try {
      render(
        <SafeAreaProvider>
          <Tooltip
            id={'test'}
            content={'content Lorem ipsum dolor sit amet'}
            position={'top'}>
            <Text>trigger text</Text>
          </Tooltip>
        </SafeAreaProvider>,
      );
    } catch (error: any) {
      expect(error.message).toBe(
        'useTooltipContext must be used within a TooltipProvider',
      );
    }
  });

  test('should show top and hide tooltip via trigger', async () => {
    // when
    const mockOnClose = jest.fn();
    const { queryByTestId } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            content={'content Lorem ipsum dolor sit amet'}
            position={'top'}
            onClose={mockOnClose}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const trigger = queryByTestId('tooltip-trigger-test');
    const childWrapper = queryByTestId('tooltip-child-wrapper-test');
    childWrapper!.parent!.instance.measure = mockChildWrapperTopMeasure;
    fireEvent(childWrapper, 'layout', mockChildWrapperOnLayoutParam);
    fireEvent.press(trigger);

    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout!, 'layout', {});

    const content = queryByTestId('tooltip-content-test');
    expect(content).not.toBeNull();

    const close = queryByTestId('tooltip-close-test');
    fireEvent.press(close!);
    expect(queryByTestId('tooltip-content-test')).toBeNull();
    expect(mockOnClose).toHaveBeenCalledWith('test');
  });

  test('should show bottom and hide tooltip via trigger', async () => {
    // when
    const mockOnClose = jest.fn();
    const { queryByTestId } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            overlay
            hole
            content={'content Lorem ipsum dolor sit amet'}
            position={'bottom'}
            onClose={mockOnClose}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const trigger = queryByTestId('tooltip-trigger-test');
    const childWrapper = queryByTestId('tooltip-child-wrapper-test');

    childWrapper!.parent!.instance.measure = mockChildWrapperTopMeasure;

    fireEvent(childWrapper, 'layout', mockChildWrapperOnLayoutParam);
    fireEvent.press(trigger);

    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout!, 'layout', {});

    const content = queryByTestId('tooltip-content-test');
    expect(content).not.toBeNull();

    const close = queryByTestId('tooltip-close-test');
    fireEvent.press(close!);
    expect(queryByTestId('tooltip-content-test')).toBeNull();
    expect(mockOnClose).toHaveBeenCalledWith('test');
  });

  test('should show bottom when there is no space at top', async () => {
    // when
    const mockOnClose = jest.fn();

    const { queryByTestId, getByTestId, getByA11yHint } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            content={'content Lorem ipsum dolor sit amet'}
            overlay
            hole
            position={'top'}
            onClose={mockOnClose}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const trigger = queryByTestId('tooltip-trigger-test');
    const childWrapper = queryByTestId('tooltip-child-wrapper-test');
    childWrapper!.parent!.instance.measure = mockChildWrapperTopMeasure;
    fireEvent(childWrapper, 'layout', mockChildWrapperOnLayoutParam);
    fireEvent.press(trigger);

    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout!, 'layout', {});

    await waitFor(() => getByTestId('tooltip-content-test'));

    const popup = queryByTestId('tooltip-test');
    popup!.parent!.instance.measure = jest.fn().mockImplementation(f => {
      // x,y,w,h,px,py
      f(0, 0, 360, 200, 32, 62);
    });
    fireEvent(popup!, 'layout', {});
    expect(getByA11yHint('bottom')).not.toBeNull();
  });

  test('should show top when there is no space at bottom', async () => {
    // when
    const mockOnClose = jest.fn();

    const { queryByTestId, getByTestId, getByA11yHint } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            content={'content Lorem ipsum dolor sit amet'}
            overlay
            hole
            position={'bottom'}
            onClose={mockOnClose}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const trigger = queryByTestId('tooltip-trigger-test');
    const childWrapper = queryByTestId('tooltip-child-wrapper-test');
    childWrapper!.parent!.instance.measure = mockChildWrapperBottomMeasure;
    fireEvent(childWrapper, 'layout', mockChildWrapperOnLayoutParam);
    fireEvent.press(trigger);

    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout!, 'layout', {});

    await waitFor(() => getByTestId('tooltip-content-test'));

    const popup = queryByTestId('tooltip-test');
    popup!.parent!.instance.measure = jest.fn().mockImplementation(f => {
      // x,y,w,h,px,py
      f(0, 0, 360, 100, 32, deviceHeight - 100 - 21); // component height, inset bottom
    });
    fireEvent(popup!, 'layout', {});
    expect(getByA11yHint('top')).not.toBeNull();
  });

  test('should render overlay and hole', async () => {
    // when
    const { queryByTestId } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            overlay
            hole
            position={undefined as any}
            content={'content Lorem ipsum dolor sit amet'}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const trigger = queryByTestId('tooltip-trigger-test');
    const childWrapper = queryByTestId('tooltip-child-wrapper-test');
    childWrapper!.parent!.instance.measure = mockChildWrapperTopMeasure;

    fireEvent(childWrapper, 'layout', mockChildWrapperOnLayoutParam);
    fireEvent.press(trigger);

    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout, 'layout', {});

    const content = queryByTestId('tooltip-content-test');
    const overlay = queryByTestId('tooltip-overlay-test');
    const hole = queryByTestId('tooltip-hole-test');
    const contour = queryByTestId('tooltip-contour-test');
    expect(content).not.toBeNull();
    expect(overlay).not.toBeNull();
    expect(hole).not.toBeNull();
    expect(contour).toBeNull();
  });

  test('should render overlay and hole and contour', async () => {
    // when
    const { queryByTestId } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            overlay
            hole
            contour
            position="bottom"
            content={'content Lorem ipsum dolor sit amet'}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const trigger = queryByTestId('tooltip-trigger-test');
    const childWrapper = queryByTestId('tooltip-child-wrapper-test');
    childWrapper!.parent!.instance.measure = mockChildWrapperTopMeasure;

    fireEvent(childWrapper, 'layout', mockChildWrapperOnLayoutParam);
    fireEvent.press(trigger);

    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout, 'layout', {});

    const content = queryByTestId('tooltip-content-test');
    const overlay = queryByTestId('tooltip-overlay-test');
    const hole = queryByTestId('tooltip-hole-test');
    const contour = queryByTestId('tooltip-contour-test');
    expect(content).not.toBeNull();
    expect(overlay).not.toBeNull();
    expect(hole).not.toBeNull();
    expect(contour).not.toBeNull();
  });

  test('should show and hide via ref', async () => {
    // given
    const ref = React.createRef<TooltipRef>();
    const mockOnClose = jest.fn();
    // when
    const { queryByTestId } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            ref={ref}
            overlay
            content={'content Lorem ipsum dolor sit amet'}
            position={'top'}
            onClose={mockOnClose}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const childWrapper = queryByTestId('tooltip-child-wrapper-test');
    childWrapper!.parent!.instance.measure = mockChildWrapperTopMeasure;
    fireEvent(childWrapper, 'layout', mockChildWrapperOnLayoutParam);

    ref.current?.show();

    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout, 'layout', {});

    const content = queryByTestId('tooltip-content-test');
    expect(content).not.toBeNull();

    ref.current?.hide();
    expect(queryByTestId('tooltip-content-test')).toBeNull();
    expect(mockOnClose).toHaveBeenCalledWith('test');
  });

  test('should clear all tooltips', async () => {
    const ref1 = React.createRef<TooltipRef>();
    const ref2 = React.createRef<TooltipRef>();

    const Comp = () => {
      const { clearAll } = useTooltipContext();

      return (
        <Box>
          <Tooltip
            id={'test1'}
            ref={ref1}
            overlay
            hole
            content={'content 1'}
            position={'top'}>
            <Text>trigger text 1</Text>
          </Tooltip>
          <Tooltip
            id={'test2'}
            ref={ref2}
            overlay
            hole
            content={'content 2'}
            position={'top'}>
            <Text>trigger text 2</Text>
          </Tooltip>

          <Button testID="clear_all_btn" onPress={clearAll} label="Clear All" />
        </Box>
      );
    };
    // when
    const { queryByTestId } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Comp />
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const childWrapper1 = queryByTestId('tooltip-child-wrapper-test1');
    const childWrapper2 = queryByTestId('tooltip-child-wrapper-test2');
    childWrapper1!.parent!.instance.measure = mockChildWrapperTopMeasure;
    childWrapper2!.parent!.instance.measure = jest
      .fn()
      .mockImplementation(f => {
        // x,y,w,h,px,py
        f(0, 0, 200, 30, 16, 64);
      });
    fireEvent(childWrapper1, 'layout', mockChildWrapperOnLayoutParam);
    fireEvent(childWrapper2, 'layout', mockChildWrapperOnLayoutParam);

    ref1.current?.show();
    ref2.current?.show();

    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test1');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout!, 'layout', {});

    const content1 = queryByTestId('tooltip-content-test1');
    const content2 = queryByTestId('tooltip-content-test2');
    expect(content1).not.toBeNull();
    expect(content2).toBeNull();

    // clear all
    const btn = queryByTestId('clear_all_btn');
    fireEvent.press(btn!);

    expect(queryByTestId('tooltip-content-test1')).toBeNull();
    expect(queryByTestId('tooltip-content-test2')).toBeNull();
  });

  test('should wait for layout ready', async () => {
    const ref = React.createRef<TooltipRef>();
    // when
    const { queryByTestId } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            ref={ref}
            overlay
            hole
            content={'content Lorem ipsum dolor sit amet'}
            position={'top'}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const childWrapper = queryByTestId('tooltip-child-wrapper-test');
    childWrapper!.parent!.instance.measure = mockChildWrapperTopMeasure;

    ref.current?.show();
    expect(queryByTestId('tooltip-content-test')).toBeNull();

    fireEvent(childWrapper, 'layout', mockChildWrapperOnLayoutParam);

    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout!, 'layout', {});

    const content = queryByTestId('tooltip-content-test');
    expect(content).not.toBeNull();
  });

  test('should hide when press overlay', async () => {
    // when
    const { queryByTestId } = render(
      <SafeAreaProvider>
        <TooltipProvider>
          <Tooltip
            id={'test'}
            overlay
            hole
            content={'content Lorem ipsum dolor sit amet'}
            position={'top'}>
            <Text>trigger text</Text>
          </Tooltip>
        </TooltipProvider>
      </SafeAreaProvider>,
    );

    // then
    const childWrapper = queryByTestId('tooltip-child-wrapper-test');
    childWrapper!.parent!.instance.measure = mockChildWrapperTopMeasure;

    fireEvent(childWrapper, 'layout', mockChildWrapperOnLayoutParam);

    const trigger = queryByTestId('tooltip-trigger-test');
    fireEvent.press(trigger);
    const overlayLayout = queryByTestId('tooltip-overlay-wrapper-test');
    overlayLayout!.parent!.instance.measure = mockOverlayMeasure;
    fireEvent(overlayLayout, 'layout', {});

    fireEvent.press(overlayLayout);
    expect(queryByTestId('tooltip-content-test')).toBeNull();
  });
});
