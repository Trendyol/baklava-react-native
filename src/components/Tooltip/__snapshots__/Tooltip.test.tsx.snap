// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render bottom correctly 1`] = `
<View
  accessibilityLabel="tooltip-trigger-test"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="tooltip-trigger-test"
>
  <View
    accessibilityLabel="tooltip-child-wrapper-test"
    accessible={true}
    onLayout={[Function]}
    style={
      [
        {},
      ]
    }
    testID="tooltip-child-wrapper-test"
  >
    <Text
      accessible={false}
      style={
        [
          {
            "color": "#273142",
            "fontFamily": "Rubik-Regular",
            "fontSize": 14,
          },
        ]
      }
    >
      trigger text
    </Text>
  </View>
</View>
`;

exports[`Tooltip should render top correctly 1`] = `
<View
  accessibilityLabel="tooltip-trigger-test"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="tooltip-trigger-test"
>
  <View
    accessibilityLabel="tooltip-child-wrapper-test"
    accessible={true}
    onLayout={[Function]}
    style={
      [
        {},
      ]
    }
    testID="tooltip-child-wrapper-test"
  >
    <Text
      accessible={false}
      style={
        [
          {
            "color": "#273142",
            "fontFamily": "Rubik-Regular",
            "fontSize": 14,
          },
        ]
      }
    >
      trigger text
    </Text>
  </View>
</View>
`;
